File: loader/hardware.c
/*
 * hardware.c - various hardware probing functionality
 *
 * Copyright (C) 1997, 1998, 1999, 2000, 2001, 2002, 2003  Red Hat, Inc.
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Author(s): Erik Troan <ewt@redhat.com>
 *            Matt Wilson <msw@redhat.com>
 *            Michael Fulbright <msf@redhat.com>
 *            Jeremy Katz <katzj@redhat.com>
 */

#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <strings.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <sys/wait.h>
#include <glib.h>

#include "loader.h"
#include "hardware.h"

/* FIXME: for turning off dma */
#include <sys/ioctl.h>
#include <linux/hdreg.h>
#include "../isys/isys.h"
#include "../isys/log.h"

/* boot flags */
extern uint64_t flags;

static int detectHardware() {
    int child, rc, status;
    int timeout = 0; /* FIXME: commandline option for this */

    fprintf(stderr, "detecting hardware...\n");
    logMessage(DEBUGLVL, "probing buses");

    if (!(child = fork())) {
        int fd = open("/dev/tty3", O_RDWR);

        dup2(fd, 0);
        dup2(fd, 1);
        dup2(fd, 2);
        close(fd);

        rc = execl("/sbin/udevadm", "udevadm", "trigger", NULL);
        _exit(1);
    }

    waitpid(child, &status, 0);
    if (!WIFEXITED(status) || (WIFEXITED(status) && WEXITSTATUS(status))) {
        rc = 1;
    } else {
        rc = 0;
    }

    fprintf(stderr, "waiting for hardware to initialize...\n");
    logMessage(DEBUGLVL, "waiting for hardware to initialize");

    if (!(child = fork())) {
        char *args[] = { "/sbin/udevadm", "settle", NULL, NULL };
        int fd = open("/dev/tty3", O_RDWR);

        dup2(fd, 0);
        dup2(fd, 1);
        dup2(fd, 2);
        close(fd);

        if (timeout) {
            checked_asprintf(&args[2], "--timeout=%d", timeout);
        }

        rc = execv("/sbin/udevadm", args);
        _exit(1);
    }

    waitpid(child, &status, 0);
    if (!WIFEXITED(status) || (WIFEXITED(status) && WEXITSTATUS(status))) {
        rc = 1;
    } else {
        rc = 0;
    }
    if (rc) {
        return LOADER_ERROR;
    }
    return LOADER_OK;
}

/* this allows us to do an early load of modules specified on the
 * command line to allow automating the load order of modules so that
 * eg, certain scsi controllers are definitely first.
 * FIXME: this syntax is likely to change in a future release
 *        but is done as a quick hack for the present.
 */
int earlyModuleLoad(int justProbe) {
    int fd, len, i;
    char buf[1024], *cmdLine;
    gint argc = 0;
    gchar **argv = NULL;
    GError *optErr = NULL;

    /* FIXME: reparsing /proc/cmdline to avoid major loader changes.  
     * should probably be done in loader.c:parseCmdline() like everything 
     * else
     */
    if ((fd = open("/proc/cmdline", O_RDONLY)) < 0) return 1;
    len = read(fd, buf, sizeof(buf) - 1);
    close(fd);
    if (len <= 0) return 1;

    buf[len] = '\0';
    cmdLine = buf;

    if (!g_shell_parse_argv(cmdLine, &argc, &argv, &optErr)) {
        g_error_free(optErr);
        return 1;
    }

    for (i=0; i < argc; i++) {
        if (!strncasecmp(argv[i], "driverload=", 11)) {
            logMessage(INFO, "loading %s early", argv[i] + 11);
            mlLoadModuleSet(argv[i] + 11);
        }
    }
    return 0;
}

int busProbe(int justProbe) {
    /* autodetect whatever we can */
    if (justProbe)
        return 0;
    return detectHardware();
}

File: pyanaconda/ui/validators/hardware.py
# The class for hardware validation.
#
# Copyright (C) 2016  Red Hat, Inc.
#
# This copyrighted material is made available to anyone wishing to use,
# modify, copy, or redistribute it subject to the terms and conditions of
# the GNU General Public License v.2, or (at your option) any later version.
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY expressed or implied, including the implied warranties of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.  You should have received a copy of the
# GNU General Public License along with this program; if not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
# source code or documentation are not subject to the GNU General Public
# License and may only be used or replicated with the express permission of
# Red Hat, Inc.
#
# Author(s):  Vendula Poncova <vponcova@redhat.com>
#

import logging

from pyanaconda.i18n import N_, _
from pyanaconda.iutil import is_unsupported_hw
from pyanaconda.product import productName
from pyanaconda.ui.validators import BaseValidator

log = logging.getLogger("anaconda")

__all__ = ["HardwareValidator"]


class HardwareValidator(BaseValidator):
    """A class to check that the hardware is supported."""

    title = N_("Hardware validation")

    def __init__(self, config):
        super(HardwareValidator, self).__init__(config)
        self._data = config.data

    def _is_valid(self):
        # Is the hardware supported?
        # pylint: disable=no-member
        return not productName.startswith("Red Hat ") \
               or not is_unsupported_hw() \
               or self._data.unsupportedhardware.unsupported_hardware

    def _get_validation_error(self):
        return _("This hardware (or a combination thereof) is not "
                 "supported by Red Hat. For more information on "
                 "supported hardware, please refer to "
                 "http://www.redhat.com/hardware.")
